Add support for SHOW CREATE SCHEMA.

lhofhansl opened the pull request:
~~Needs #3066 (when that is merged, I'll rebase on master)~~

~~WIP!
No tests. Basically works.~~

Ready for a review.

lhofhansl commented on presto-main/src/main/java/io/prestosql/sql/rewrite/ShowQueriesRewrite.java:
```diff
@@ -486,6 +495,33 @@ protected Node visitShowCreate(ShowCreate node, Void context)
+                CatalogSchemaName schemaName = createCatalogSchemaName(session, node, Optional.of(node.getName()));
+                Optional<ConnectorSchemaDefinition> schema = metadata.getSchema(session, schemaName);
+
+                // TODO: at present everyone can see schema metadata
```
This will need discussion. If only an owner or admin can do this, it's kinda pointless. On the other hand information can be leaked this way: Where the schema is located or that the assigned USER or ROLE exists.

Maybe limiting this to Owner/Admin is fine.

lhofhansl commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorSchemaDefinition.java:
```diff
@@ -0,0 +1,50 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class ConnectorSchemaDefinition
```
This would also need discussion. I'm using it only when retrieving schema info, not when creating a schema (to avoid changing many dozens of interfaces/classes)

lhofhansl commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -521,6 +522,15 @@ default void dropView(ConnectorSession session, SchemaTableName viewName)
+    default Optional<ConnectorSchemaDefinition> getSchema(ConnectorSession session, CatalogSchemaName schemaName)
+    {
+        // Only the Hive Connector supports this. This default simplifies all other connectors.
+        return Optional.of(new ConnectorSchemaDefinition(Collections.emptyMap(), Optional.empty()));
```
See the comment... With this we can avoid adding a new method to each and every connector.

findepi commented:
@lhofhansl very cool. Will the SHOW CREATE SCHEMA also show schema location (in case of hive connector), or more generally, schema properties?

lhofhansl commented:
@findepi Yep, it shows the location as well.
That is the only property that exists for now (for Hive at least), but I kept it general so that connectors can add any property they want, and it'd show in the WITH clause).
Actually to make it really general I should bite the bullet and change the signature of createSchema to accept a ConnectorSchemaDefinition rather than individual arguments.

lhofhansl commented:
@kokosing I agree, I will make that so. Just wanted to get agreement before I change a million classes :)

lhofhansl commented:
add checkCanShowCreateSchema... Only an admin or schema owner can see a schema's definition.
Test manually... Next: Tests!

lhofhansl commented:
Added tests.

lhofhansl commented:
Thoughts? :)

kokosing commented:
This in general looks good to me, but I want first merge: #3066 So then I will focus on it.

lhofhansl commented:
Thanks and agreed.

lhofhansl commented:
With #3066 merged, rebased directly onto master. @kokosing

lhofhansl commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -713,6 +713,55 @@ public void testSchemaAuthorization()
+                "CREATE SCHEMA %s.test_show_create_schema\n" +
+                "AUTHORIZATION USER hive\n" +
+                "WITH \\(\n" +
+                "   location = '.*test_show_create_schema'\n" +
```
Hmm... I wouldn't know how to distinguish this. The location *is* set in the metadata, either a default or a given value. I'll dig into this.

lhofhansl replied:
Schema is handled differently. The Hivemetastore sets the default, so it's always set. Without some bigger effort I cannot hide it if it is the Hive default.

Table location, on the other hand, is shown when the table's type is "external table", otherwise it isn't even set.

lhofhansl replied:
Found a way by using the metastore's get_config_value. I also implemented that for the FileHiveMetastore by supporting the warehouse config there.

lhofhansl replied:
Now of course I cannot test showing the location anymore. Adding this:
```
        // now test with a custom schema location
        File tempDir = createTempDir();
        assertUpdate(admin, format("CREATE SCHEMA test_schema_with_location WITH (location = '%s')", tempDir));

        createSchemaSql = format("" +
                "CREATE SCHEMA %s.test_schema_with_location\n" +
                "AUTHORIZATION USER hive\n" +
                "WITH \\(\n" +
                "   location = '%s'\n" +
                "\\)",
                getSession().getCatalog().get(),
                tempDir);

        actualResult = getOnlyElement(computeActual(admin, "SHOW CREATE SCHEMA test_schema_with_location").getOnlyColumnAsSet()).toString();
        assertEquals(actualResult, createSchemaSql);
        assertUpdate(user, "DROP SCHEMA test_schema_with_location");
```
Does not work, since the metastore used for tests does not allow setting schema,

lhofhansl commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1732,6 +1734,26 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
     }
 
+    @Override
+    public Optional<ConnectorSchemaDefinition> getSchema(ConnectorSession session, CatalogSchemaName schemaName)
```
I was trying to avoid polluting the interface even more, and the owner is kinda part of the definition of a schema. I don't feel strongly, though.

lhofhansl replied:
It needs to be filterSchema to filter out the information_schema and the "sys" schema. Schema existence is checked higher in the call stack and in addition the Database is an Optional.

lhofhansl commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1732,6 +1734,26 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
+
+                    // this is the only current properties for Hive
+                    Optional<String> dbLocation = db.getLocation();
+                    dbLocation.ifPresent(location -> properties.put(HiveSchemaProperties.LOCATION_PROPERTY, location));
```
Not sure how I would that. Each property has a different source. So while I could loop, each value (like location here) has a different way to get the value, unless we generalize that and match it up with the Hive properties.

lhofhansl commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -713,6 +713,55 @@ public void testSchemaAuthorization()
+                getSession().getCatalog().get());
+
+        String actualResult = getOnlyElement(computeActual(admin, "SHOW CREATE SCHEMA test_show_create_schema").getOnlyColumnAsSet()).toString();
+        assertTrue(actualResult.matches(createSchemaSql), "Expected [" + createSchemaSql + "] but found [" + actualResult + "]");
```
This is doing regex pattern matching, so assertEquals won't work.

lhofhansl commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -713,6 +713,55 @@ public void testSchemaAuthorization()
+                getSession().getCatalog().get());
+
+        actualResult = getOnlyElement(computeActual(admin, "SHOW CREATE SCHEMA test_show_create_schema").getOnlyColumnAsSet()).toString();
+        assertTrue(actualResult.matches(createSchemaSql), "Expected [" + createSchemaSql + "] but found [" + actualResult + "]");
```
Same here...

kokosing replied:
Please introduce dedicated `assertMatch` to do regex based assertions.

kokosing replied:
Please see https://github.com/prestosql/presto/pull/3323/files#diff-cdc872cbe68e8a7497f8d7f189d58379R159, it looks much better

lhofhansl commented:
@kokosing Thanks for having a look! Pushed an update. Also see my comments above.

lhofhansl commented:
What do you think @kokosing ? I hid the LOCATION unless it differs from the Metastore's default.
(which means I can't test that, since the Hive tests use the file based metastore, which does not allow for setting a schema location)

findepi commented:
@kokosing @electrum @martint are you OK with removing `syntax-needs-review`?

kokosing commented:
>  I hid the LOCATION unless it differs from the Metastore's default.

This is good, thank you!

>  file based metastore, which does not allow for setting a schema location

maybe we could fix that in file based metastore?

> are you OK with removing syntax-needs-review?

I am OK with this.

lhofhansl commented:
>> file based metastore, which does not allow for setting a schema location

> maybe we could fix that in file based metastore?

Lemme see. Not sure we _want_ that. Perhaps do that in another PR.

lhofhansl commented:
Hmm... I forgot that Optional.or(...) only exists since Java 9.

lhofhansl commented:
I pushed a new rebase (no other changes), but due to the github outage it does not seem to show up here - even though the branch is actually updated.

lhofhansl commented:
Rebased again... Seems fine now.

lhofhansl commented:
Looks like the tests are failing due to lack of disk-space.

findepi commented:
> Looks like the tests are failing due to lack of disk-space.

see discussion on our [slack](https://prestosql.io/slack.html) there: https://prestosql.slack.com/archives/CP1MUNEUX/p1585838567075000
we're working on this.

lhofhansl commented:
Ahh... That looks much better. Thanks for fixing it!!

lhofhansl commented:
Obligatory `<ping>` :)
Seems to pass all tests, has tests, should be good.

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1732,6 +1734,36 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
+            // this is currently the only property for Hive
+            // (only shown when it's not the default location)
+            if (location.isPresent() && !location.get().startsWith(warehouse)) {
+                return Collections.singletonMap(HiveSchemaProperties.LOCATION_PROPERTY, location.get());
```
`ImmutableMap.of()`

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1732,6 +1734,36 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
+            Optional<String> location = db.get().getLocation();
+            // this is currently the only property for Hive
+            // (only shown when it's not the default location)
+            if (location.isPresent() && !location.get().startsWith(warehouse)) {
```
I think this condition is not enough. Is it possible that one could create external table under `warehouse` location? Like pointing to data of another managed table? I do think there is a check for that in Hive and this allowed. We need to verify full path, not only prefix.

Please add test for that.

lhofhansl replied:
Hmm... I was debating this as well. I'd have to hardcode a convention then, name that the default name for any database is `<datewarehouse.dir>/<databasename>.db`
It's not clear what it means if somebody now manually creates a directory inside the `datawarehouse.dir` and then places a schema (not table, there are no external schemas) in there.
I can certain verify that the path is *exactly* `<datewarehouse.dir>/<databasename>.db` and show it otherwise.

lhofhansl replied:
The default is `<warehouse.dir>/<schemaname>.db`, but any path is allowed (without ending in .db). So I think we should verify `<warehouse.dir>/<schemaname>` exactly.

But people can vary the "extension" too. The file based hive metastore has no extensions.
So if you name your location `<warehouse.dir>/<schemaname>.mydb` should I show the location or not?

Also looks like there are no further checks. I can create a schema in `/a/b/c` and also in `/a/b`, then create a table `t` in `/a/b`. Now we have a schema `c` and table `t` in the same directory.

@kokosing

lhofhansl replied:
BTW. Hive always shows the location whether it's default or not.
Perhaps it's better to follow that?

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1732,6 +1734,36 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
+    public Map<String, Object> getSchemaProperties(ConnectorSession session, CatalogSchemaName schemaName)
+    {
+        if (!filterSchema(schemaName.getSchemaName())) {
+            return Collections.emptyMap();
```
`ImmtableMap.of()`

This is an unwritten convention.

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1732,6 +1734,36 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
+            }
+        }
+
+        return Collections.emptyMap();
```
`ImmtableMap.of()`

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1732,6 +1734,36 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
+    @Override
+    public Optional<PrestoPrincipal> getSchemaOwner(ConnectorSession session, CatalogSchemaName schemaName)
+    {
+        if (!filterSchema(schemaName.getSchemaName())) {
```
This is not responsibility of connector to filter schemas. It should be handled somewhere in `ShowQueriesRewrite`

Also if user is allowed to do show create schema then we don't need to filter schema here.

lhofhansl replied:
This is for the information_schema and for Hive's sys schema.
(see `listSchemas`, `listTables` methods, etc.)

lhofhansl replied:
That is to say: information_schema and sys have no owner as such, and they also do not have any properties.

kokosing replied:
Please do not use `filterSchema` but just make sure that requested schema is one of these. And if it is the case I think we should throw an exception.

@findepi I think you are more familiar with sys schema. Are ok with my suggestion?

findepi replied:
See https://github.com/prestosql/presto/pull/3099#discussion_r406612194

`getSchemaOwner` should look similar to getSchemaProperties

(btw, i wasn't following -- did you consider combining them into single getSchemaMetadata operation?)

kokosing commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -716,6 +756,66 @@ public void testSchemaAuthorization()
+        assertEquals(actualResult, createSchemaSql);
+
+        // test that LOCATION is shown when it does not match the metastore's default
+        testingMetastore.overrideConfig(WAREHOUSE.getVarname(), "x");
```
You should use io.prestosql.testing.AbstractTestQueryFramework#executeExclusively

kokosing commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -716,6 +756,66 @@ public void testSchemaAuthorization()
+        actualResult = getOnlyElement(computeActual(admin, "SHOW CREATE SCHEMA test_show_create_schema").getOnlyColumnAsSet()).toString();
+        assertTrue(actualResult.matches(createSchemaSql), "Expected [" + createSchemaSql + "] but found [" + actualResult + "]");
+
+        testingMetastore.overrideConfig(WAREHOUSE.getVarname(), null);
```
provide `reset()` method that would reset all the custom configuration

kokosing commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -960,6 +961,34 @@ public void finishDelete(Session session, TableHandle tableHandle, Collection<Sl
+    @Override
+    public Map<String, Object> getSchemaProperties(Session session, CatalogSchemaName schemaName)
+    {
+        if (schemaExists(session, schemaName)) {
```
I don't like this optionality. When you ask for properties of not existing object this method should fail. Returning empty map suggests that object exists but has no properties.

Same to the method below.

lhofhansl replied:
Hmm... I was following other examples such as getView and getTable. But these *are* different.

kokosing replied:
This is more like `io.prestosql.metadata.MetadataManager#getTableProperties`

lhofhansl replied:
Added an error check to both methods.

kokosing commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -960,6 +961,34 @@ public void finishDelete(Session session, TableHandle tableHandle, Collection<Sl
+            ConnectorSession connectorSession = session.toConnectorSession(catalogName);
+            return metadata.getSchemaProperties(connectorSession, schemaName);
+        }
+        return Collections.emptyMap();
```
ImmutableMap

kokosing commented on presto-main/src/main/java/io/prestosql/sql/rewrite/ShowQueriesRewrite.java:
```diff
@@ -486,7 +494,31 @@ protected Node visitShowCreate(ShowCreate node, Void context)
+                CatalogSchemaName schemaName = createCatalogSchemaName(session, node, Optional.of(node.getName()));
+                accessControl.checkCanShowCreateSchema(session.toSecurityContext(), schemaName);
+
+                if (metadata.schemaExists(session, schemaName)) {
```
invert condition. Then `else` block is not needed.

lhofhansl commented:
@kokosing Thanks for looking. I like the detailed code reviews! 
And sorry if I seem pushy - I'm just making sure that there's no circular waiting.

lhofhansl commented:
@kokosing Pushed an update the addresses all comments except these too:
1. FilterSchema in HiveMetadata. This just filters `information_schema` and `sys`, which neither have been created by a user, nor they have properties or an owner, so an empty return should be appropriate
2. Not showing the LOCATION when a prefix matches the `warehouse.dir`. I think perhaps we should just remove that part. Hive always shows the LOCATION (see the comment above as well), and dealing with all possibilities (varying pre- and suffix) is error prone.

lhofhansl commented:
Rebased and removed hiding the LOCATION. The location is different from external tables in that it's not optional. The metastore has to set a real value for each schema, otherwise changing the warehouse.dir config would render all schemas invalid.
External tables do not have the same problem since a schema's location cannot be altered, so an optional external location is OK.

So I now think it's better to always show the location (like Hive does as stated above)

@kokosing (no pressure, I know you guys are busy ATM)

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1738,6 +1740,41 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
+        Optional<Database> db = metastore.getDatabase(schemaName.getSchemaName());
+        if (db.isPresent()) {
+            Optional<String> location = db.get().getLocation();
+            // this is currently the only property for Hive
```
I would you to use `io.prestosql.plugin.hive.HiveSchemaProperties#SCHEMA_PROPERTIES`. So whenever someone adds new schema property they don't need to do any changes here.

lhofhansl replied:
Not sure what you mean. I would still need to add custom code here for each property to retrieved. For example the location is read from the Database object. Other properties have other ways to be retrieved.

Can you explain what you mean? Apologies if I'm missing something obvious.
Do you mean add a method to HiveSchemaProperties that is being passed a Database object? Or more like HiveTableProperties?

lhofhansl replied:
I added method to HiveSchemaProperties. Lemme know what you think.

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1738,6 +1740,41 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
+    public Map<String, Object> getSchemaProperties(ConnectorSession session, CatalogSchemaName schemaName)
+    {
+        if (!filterSchema(schemaName.getSchemaName())) {
+            // "information_schema" and "sys" have no properties
```
Please do not use `filterSchema` but just make sure that requested schema is one of these. And if it is the case I think we should throw an exception.

@findepi I think you are more familiar with `sys` schema. Are ok with my suggestion?

lhofhansl replied:
That's precisely what `HiveMetadata.filterSchema` does, though. It's a static method that checks for these schemas (I suppose it's badly named).
I can throw an exception instead, but I'd just do the same check with extra code.

lhofhansl replied:
Should I add a new error code to `HiveErrorCode`?

lhofhansl replied:
Have look. Throwing NOT_SUPPORTED exception now.

findepi replied:
Can engine call `getSchemaProperties` without calling `schemaExists` first? (Does engine guarantee that?)

* yes => sys needs to be filtered out; and we should return NOT_FOUND (because we pretend the "sys" schema does not exist, as indicated by `false` from `schemaExists`)
* no => `checkState` would be more appropriate: `checkState(filterSchema(..), "Schema is not accessible: %s", schemaName)`

lhofhansl replied:
This is indeed only called from `ShowQueriesRewrite,Visitor.visitShowCreate(...)`, which does check `schemaExists` first and throws SCHEMA_NOT_FOUND from there. (Although it's not guaranteed that nobody will add such a call in the future.)

Added checkState bit...

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -224,6 +225,15 @@ public synchronized void setDatabaseOwner(HiveIdentity identity, String database
     }
 
+    @Override
+    public Optional<String> getConfigValue(String name)
```
Do you still need it?

lhofhansl replied:
I thought it might not hurt to have it . Happy to remove it.

lhofhansl replied:
I'll remove it.

kokosing commented on presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:
```diff
@@ -115,6 +116,11 @@ protected Object computeScalar(@Language("SQL") String sql)
 
+    protected void assertMatch(String actual, String regex)
+    {
+        assertTrue(actual.matches(regex), "[" + actual + "] does not match [" + regex + "]");
```
the other day I left a comment to use assertj, it has some native support for regex matching. If you apply that comment then I would expect this code become very fluent and so we might have no longer a need for this method (we could just inline it).

assertMatch -> assertMatches

kokosing commented on presto-testing/src/main/java/io/prestosql/testing/AbstractTestQueryFramework.java:
```diff
@@ -115,6 +116,11 @@ protected Object computeScalar(@Language("SQL") String sql)
         return computeActual(sql).getOnlyValue();
     }
 
+    protected void assertMatch(String actual, String regex)
```
static. It does not belong to AbstractTestQueryFramework, but rather to something like `PrestoAssert`.

kokosing commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -521,6 +522,22 @@ default void dropView(ConnectorSession session, SchemaTableName viewName)
+     */
+    default Map<String, Object> getSchemaProperties(ConnectorSession session, CatalogSchemaName schemaName)
+    {
+        return Collections.emptyMap();
```
I think it should throw by default.

lhofhansl replied:
Hmm... Should `SHOW CREATE SCHEMA` fail on every connector except the Hive connector?

lhofhansl replied:
I suppose any connector that does not support schemas should - and safely - throw an exception.

kokosing commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:
```diff
@@ -521,6 +522,22 @@ default void dropView(ConnectorSession session, SchemaTableName viewName)
+     */
+    default Optional<PrestoPrincipal> getSchemaOwner(ConnectorSession session, CatalogSchemaName schemaName)
+    {
+        return Optional.empty();
```
I think it should throw by default.

lhofhansl replied:
Same here, should `SHOW CREATE SCHEMA` fail on every connector except the Hive connector?

lhofhansl commented:
Rebased. @kokosing FYI.

kokosing commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadata.java:
```diff
@@ -82,6 +83,18 @@ public boolean schemaExists(ConnectorSession session, String schemaName)
+    @Override
+    public Optional<PrestoPrincipal> getSchemaOwner(ConnectorSession session, CatalogSchemaName schemaName)
+    {
+        return Optional.empty();
```
Is this correct? I guess RDBMS might have a notion of schema owner. Maybe we should leave it unimplemented?

lhofhansl replied:
Throw UnsupportedException?
What I did was to have an implementation that does not fail for every connector that supports `CREATE SCHEMA`. Wouldn't it be weird to have `SHOW CREATE SCHEMA` fail with an exception here?
No problem removing the implementation here, we have to just be aware of the implications.

lhofhansl replied:
I think we might add properties for schemas in the Jdbc Connector in the future. So getSchema  Owner (IMHO) should not fail.

kokosing replied:
> Wouldn't it be weird to have SHOW CREATE SCHEMA fail with an exception here?

If it is better to fail than return incorrect data. In other words, JDBC connectors is a pandora box let's not open it for now.

lhofhansl replied:
Ok... Lemme do that.

lhofhansl replied:
Also... I assume the same holds for KuduMetadata, BlackHoleMetaData, MemoryMetadata, PhoenixMetadata?

The logic I followed was that if a connector has `CREATE SCHEMA` it should also not fail when doing `SHOW CREATE SCHEMA` even when there's no info to show.

Unless I hear otherwise from you I will remove all no-op implementation of these methods and have all those connectors fail when issuing `SHOW CREATE SCHEMA`.

kokosing replied:
BlackHoleMetaData are trivial MemoryMetadata so you could implement that method for them.

I would skip rest for now.

lhofhansl replied:
Oh missed this. I will add this back for BlockHoleMetadata and MemoryMetadata.

kokosing commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -716,6 +716,56 @@ public void testSchemaAuthorization()
     }
 
+    @Test
+    public void testShowCreateSchema()
```
We need also add simple test to `io.prestosql.testing.AbstractTestIntegrationSmokeTest`. That file is used to test connector implementations, you change connector SPI so we need tests for other connectors. It can be very simple, without any access control. Probably for most of the connectors it will be skipped.

lhofhansl replied:
I'm not sure what I would test.

It's kinda like `TestHiveIntegrationSmokeTest.testSchemaOperations`, which _only_ exists in `TestHiveIntegrationSmokeTest`, because it doesn't make sense for many other connectors, and there's no equivalent for this test in the other connector tests.

Do you have a suggestions about what I can test generally?

kokosing replied:
I was thinking to add simple `SHOW CREATE SCHEMA tpch`. If connector does not support this, then we could disable that test for this connector, but it will be known that there is a feature that given connector does not support.

lhofhansl replied:
I see. Makes sense.
Should probably pull `TestHiveIntegrationSmokeTest.testSchemaOperations` up to `AbstractTestIntegrationSmokeTest` as well then.

kokosing replied:
Makes sense

kokosing commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -957,6 +957,34 @@ public void finishDelete(Session session, TableHandle tableHandle, Collection<Sl
+            ConnectorSession connectorSession = session.toConnectorSession(catalogName);
+            return metadata.getSchemaProperties(connectorSession, schemaName);
+        }
+        return ImmutableMap.of();
```
I think we should fail if schema does not exists

kokosing commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -957,6 +957,34 @@ public void finishDelete(Session session, TableHandle tableHandle, Collection<Sl
+            ConnectorSession connectorSession = session.toConnectorSession(catalogName);
+            return metadata.getSchemaOwner(connectorSession, schemaName);
+        }
+        return Optional.empty();
```
I think we should fail if schema does not exists

kokosing commented on presto-main/src/main/java/io/prestosql/sql/rewrite/ShowQueriesRewrite.java:
```diff
@@ -486,7 +494,30 @@ protected Node visitShowCreate(ShowCreate node, Void context)
+                CatalogSchemaName schemaName = createCatalogSchemaName(session, node, Optional.of(node.getName()));
+                accessControl.checkCanShowCreateSchema(session.toSecurityContext(), schemaName);
+
+                if (!metadata.schemaExists(session, schemaName)) {
```
I think we should check existence of object before access control

lhofhansl replied:
See `visitShowTables`. It checks access first. It feels also generally better to check access first, otherwise you can "poke" the system (by trying schema names) and gain information you should not access to. Currently that might to be an issue with schema, but we might tighten visibility in the future.

lhofhansl replied:
Wait... No, you're right. Show Tables is different. I'll change it.

lhofhansl commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1748,6 +1751,32 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
+    @Override
+    public Map<String, Object> getSchemaProperties(ConnectorSession session, CatalogSchemaName schemaName)
+    {
+        checkState(filterSchema(schemaName.getSchemaName()), "Schema is not accessible: %s", schemaName);
```
We have discussed this above :) See @findepi 's comment. It's not access control.

This is only for the `information_schema` and `sys`. And those are not available for properties or ownership. These should never be triggered, since all callers check for schemaExists first, it's just a safety consistency check in case someone adds a call to this later.

kokosing replied:
I see, thanks!

lhofhansl commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1748,6 +1751,32 @@ public void dropView(ConnectorSession session, SchemaTableName viewName)
+    @Override
+    public Optional<PrestoPrincipal> getSchemaOwner(ConnectorSession session, CatalogSchemaName schemaName)
+    {
+        checkState(filterSchema(schemaName.getSchemaName()), "Schema is not accessible: %s", schemaName);
```
See above :)

lhofhansl commented:
Pushed another update: (1) Removed getConfigValue from FileHiveMetastore, (2) check for schema existence before access control, and (3) have the methods in MetadataManager fail if the schema does not exist.

For the rest, please see the inline comments.

Thanks for looking again @kokosing

lhofhansl commented:
Removed the empty implementations of `getSchemaOwner` and `getSchemeProperties`. All connectors other than Hive and Iceberg will fail `SHOW CREATE SCHEMA` now.

I will think more on the general tests. (I would like to file a separate PR and full `testSchemaOperation` into `AbstractTestIntegrationSmokeTest`. Since many connector support CREATE/DROP SCHEMA)

lhofhansl commented:
Added a simple test for `SHOW CREATE SCHEMA` to the Iceberg tests.

lhofhansl commented:
Added default methods back for BlockHoleMetadata and MemoryMetadata.

lhofhansl commented:
Rebased. @kokosing

kokosing commented on presto-main/src/test/java/io/prestosql/security/TestAccessControlManager.java:
```diff
@@ -490,6 +495,12 @@ public void checkCanShowCreateTable(ConnectorSecurityContext context, SchemaTabl
+        @Override
+        public void checkCanShowCreateSchema(ConnectorSecurityContext context, String schemaName)
+        {
+            throw new UnsupportedOperationException();
```
FYI https://github.com/prestosql/presto/pull/3484

kokosing commented on presto-parser/src/main/java/io/prestosql/sql/SqlFormatter.java:
```diff
@@ -823,7 +823,7 @@ protected Void visitCreateSchema(CreateSchema node, Integer context)
             builder.append(formatName(node.getSchemaName()));
             if (node.getPrincipal().isPresent()) {
-                builder.append(" AUTHORIZATION ")
+                builder.append("\nAUTHORIZATION ")
```
separate commit?

lhofhansl replied:
Let's leave it this was if we can. Matches all the tests now.

kokosing commented on presto-parser/src/main/java/io/prestosql/sql/tree/ShowCreate.java:
```diff
@@ -28,7 +28,8 @@
         TABLE,
-        VIEW
+        VIEW,
+        SCHEMA
```
add:
```
SCHEMA,
/**/;
```
That way you can add new entries with changing only single line.

lhofhansl replied:
Done.

kokosing commented on presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorAccessControl.java:
```diff
@@ -122,6 +122,13 @@ default void checkCanShowSchemas(ConnectorSecurityContext context)
+     *
+     * @throws io.prestosql.spi.security.AccessDeniedException if not allowed
+     */
+    void checkCanShowCreateSchema(ConnectorSecurityContext context, String schemaName);
```
make it deny by default, see https://github.com/prestosql/presto/pull/3484

lhofhansl replied:
Done.

kokosing commented:
> @kokosing @electrum @martint are you OK with removing syntax-needs-review?

I am ok, @electrum @martint are you OK with this?

lhofhansl commented on presto-main/src/test/java/io/prestosql/metadata/AbstractMockMetadata.java:
```diff
@@ -387,6 +387,18 @@ public void finishDelete(Session session, TableHandle tableHandle, Collection<Sl
         throw new UnsupportedOperationException();
     }
 
+    @Override
```
I do not think that is true. This implements `Metadata` not `ConnectorMetadata`.

lhofhansl commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -957,6 +958,34 @@ public void finishDelete(Session session, TableHandle tableHandle, Collection<Sl
+    @Override
+    public Map<String, Object> getSchemaProperties(Session session, CatalogSchemaName schemaName)
+    {
+        if (schemaExists(session, schemaName)) {
```
Done.

lhofhansl commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -957,6 +958,34 @@ public void finishDelete(Session session, TableHandle tableHandle, Collection<Sl
+    @Override
+    public Optional<PrestoPrincipal> getSchemaOwner(Session session, CatalogSchemaName schemaName)
+    {
+        if (schemaExists(session, schemaName)) {
```
Done.

lhofhansl commented on presto-main/src/test/java/io/prestosql/security/TestAccessControlManager.java:
```diff
@@ -225,6 +225,11 @@ public void checkCanSetSystemSessionProperty(SystemSecurityContext context, Stri
             }
 
+            @Override
+            public void checkCanShowCreateSchema(ConnectorSecurityContext context, String schemaName)
```
We don't :)
I removed it.

lhofhansl commented:
I'll rebase for the conflicts...

lhofhansl commented:
Rebased... Had to add to back an import for denyShowCreateTable back to ConnectorAccessControl to get it to compile - I guess that was removed in another commit.

lhofhansl commented:
Pushed again to get another test run. Looks like a bunch of the test hung. (and some Kudu problems)

martint commented:
> are you OK with removing syntax-needs-review?

Yes, syntax looks good. It's consistent with how we do `SHOW CREATE` for other entities.

kokosing commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java:
```diff
@@ -164,6 +164,11 @@ public SemiTransactionalHiveMetastore(
         return delegate.getAllDatabases();
     }
 
+    public Optional<String> getConfigValue(String name)
```
Please remove, it is not used.

lhofhansl commented:
@kokosing Rebased and automation passes.

kokosing commented:
Merged, thanks!

kokosing commented:
BTW. Well done!

lhofhansl commented:
Awesome. Thank you. And thanks for the thorough reviews!
